### PR TITLE
Iterator extensions

### PR DESCRIPTION
**Pull request by pthariensflame**
Adds an `Extend` `impl` and a variety of `Iterator` method overrides (like the existing `fold` one).

### CODE REVIEW

**Comment by pthariensflame** [comment]
It should be noted that there were a few `Iterator` methods (`rposition` and `unzip`) that I tried to override but that the Rust compiler complained about, so I left those out of this PR.

**Comment by bluss** [comment]
Thanks for working on Either. It's hard to argue against forwarding all the methods, I guess, though it's sad that it should be needed.

What's the motivation for the Extend impl? It seems kind of random.

**Comment by pthariensflame** [comment]
It just seemed natural, while I was working with the `std::iter` module.  I would have implemented `IntoIterator` as well, but the comment I left in its place explains why that didn't go through.

File: src/lib.rs [hunk 1]
```diff
@@ -317,10 +329,142 @@ impl<L, R> Iterator for Either<L, R>
     }
 
     fn fold<Acc, G>(self, init: Acc, f: G) -> Acc
-        where G: FnMut(Acc, Self::Item) -> Acc,
+        where G: FnMut(Acc, L::Item) -> Acc,
```
Comment (bluss) [inline src/lib.rs]: For documentation's sake, I'd use `Self::Item` all the way through. Using `L::Item` is shorter, but could be confusing.

Comment (pthariensflame) [inline src/lib.rs]: Ok!  I can make that change.

Comment (bluss) [inline src/lib.rs]: ok. Sorry I've let this wait, I can merge after that.

Comment (bluss) [inline src/lib.rs]: Remove the IntoIterator comment if you can too.

Comment (pthariensflame) [inline src/lib.rs]: Gotcha.

**Comment by pthariensflame** [comment]
@bluss Ready!

**Comment by pthariensflame** [comment]
@bluss Thanks!

**Pull request merged**